### PR TITLE
Fancier IPython-like eval command

### PR DESCRIPTION
**Pull request by martmists**
- Supports Embeds
- Pretty-print
- Shortens if too long
- Supports catching stdout
- Supports async code
- Automatically awaits awaitables if they are the return value
- Some other misc changes

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -10,7 +10,9 @@ script:
...
   "curl $AUTODEPLOY_WEBHOOK -H \"token: $AUTODEPLOY_TOKEN\""
 
+cache: pip
+
 notifications:
   email:
     on_success: never
-    on_failure: never
\ No newline at end of file
+    on_failure: never
```
Comment (gdude2002) [inline .travis.yml]: This file is out of scope of the PR, please remove it.

Comment (martmists) [reply .travis.yml]: I don't even know where that line came from, probably my auto-travis-lint?

File: bot/cogs/eval.py [hunk 2]
```diff
@@ -12,66 +19,153 @@
...
-            if code.startswith("```python"):
-                code = code[9:-3]
-            elif code.startswith("```py"):
-                code = code[5:-3]
             else:
-                code = code[3:-3]
-        elif code.startswith("`") and code.endswith("`"):
-            code = code[1:-1]
+                # Indent the 3 dots correctly
+                s = (f"{{:<{len(str(self.ln))+2}}}...: ").format("")
```
Comment (gdude2002) [inline bot/cogs/eval.py]: Can you comment this better? It's very unclear what this would do for a beginner

File: bot/cogs/eval.py [hunk 3]
```diff
@@ -12,66 +19,153 @@
...
+            "contextlib": contextlib
+        }
+
+        self.env.update(env)
+
+        # Ignore this shitcode, it works
+        _code = """
+async def func():
+    try:
+        with contextlib.redirect_stdout(self.stdout):
```
Comment (gdude2002) [inline bot/cogs/eval.py]: Will do this mess up printing stuff to stdout from the rest of the bot while the eval is being processed?

Comment (martmists) [reply bot/cogs/eval.py]: Contextlib *should* handle that. I've personally never had such issue.

File: bot/cogs/eval.py [hunk 4]
```diff
@@ -50,8 +50,23 @@ def _format(self, inp, out):
...
+                # But if it's 
+                # In [XX]:
+                #    ...:
+                # 
+                # You can see it doesn't look right.
+                # This code simply indents the dots
+                # far enough to align them.
+                # we first `str()` the line number
+                # then we get the length
+                # and do a simple {:<LENGHT}
```
Comment (gdude2002) [inline bot/cogs/eval.py]: Typo! \o/

**Comment by gdude2002** [comment]
This is from one of your other bots originally, right? What's the license on that bot?

**Comment by gdude2002** [comment]
```
./bot/cogs/eval.py:57:18: W291 trailing whitespace
./bot/cogs/eval.py:58:30: W291 trailing whitespace
./bot/cogs/eval.py:6:1: I100 Import statements are in the wrong order. import pprint should be before from io
./bot/cogs/eval.py:61:18: W291 trailing whitespace
```

**Comment by martmists** [comment]
The license on said bot makes contributors unable to share it without permission, and I'm giving permission by adding it myself.

I will fix those issues in a second.

**Comment by gdude2002** [comment]
Alright, that's fine.

Really the only other thing I'd like to address in this PR is readability - there's lots of overly-short variable names, one-liners and somewhat opaque string concatenations.

We'd prefer if you could deal with that, but we can otherwise try to do it ourselves.

File: bot/cogs/eval.py [hunk 5]
```diff
@@ -12,66 +19,153 @@
...
+        # Erase temp input we made
+        if inp.startswith("_ = "):
+            inp = inp[4:]
 
-        code = string.strip()
+        lines = [l for l in inp.split("\n") if l.strip()]
+        if len(lines) != 1:
+            lines += [""]
+
+        # Create the inpit dialog
```
[the remark this answers is not in the document]
Comment (martmists) [reply bot/cogs/eval.py]: What?

File: bot/cogs/eval.py [hunk 6]
```diff
@@ -12,66 +19,153 @@
...
+                # Indent the 3 dots correctly
+                s = (f"{{:<{len(str(self.ln))+2}}}...: ").format("")
 
-        code = code.strip().strip("\n")
-        io = StringIO()
+            if i == len(lines) - 2:
+                if line.startswith("return"):
+                    line = line[6:].strip()
+
+            res += s + line + "\n"
```
[the remark this answers is not in the document]
Comment (martmists) [reply bot/cogs/eval.py]: Using f-strings or str.format here would seem unneeded to me.

File: bot/cogs/eval.py [hunk 7]
```diff
@@ -12,66 +19,153 @@
...
-                out_message = (
-                    f"{out_message}\n\n"
-                    f"**Returned**\n```py\n{repr(rvalue)}\n```"
-                )
-
-            await ctx.send(out_message)
+            if (isinstance(out, str) and
+                    out.startswith("Traceback (most recent call last):\n")):
+                # Leave out the traceback message
+                out = "\n" + "\n".join(out.split("\n")[1:])
```
Comment (triacle) [inline bot/cogs/eval.py]: it's got string concatenation, a .split nested inside a .join, and a container slice, all in one line. it's _too dense_. it'd be better to split this into more lines if it can't be simplified.

File: bot/cogs/eval.py [hunk 8]
```diff
@@ -12,66 +19,153 @@
...
+
+            pretty = (pprint.pformat(out, compact=True, width=60)
+                      if not isinstance(out, str) else str(out))
```
Comment (triacle) [inline bot/cogs/eval.py]: readability suffers because of the ternary - instead of reading the logic in semantic order (if this then that), the reader has to parse it as (that if this). And why do we need to cast `out` to a str when the isinstance clearly establishes that it is already a str?

the far simpler

```
if not isinstance(out, str):
    pretty = pprint.pformat(out, compact=True, width=60)
else:
    pretty = out
```

would be better.

**Comment by triacle** [comment]
for the record, I've left comments for absolutely all my readability concerns now. if they were all adressed, I'd approve.

**Comment by gdude2002** [comment]
Has there been any movement on this?

**Comment by martmists** [comment]
Not yet, it's been a busy week so far. I'll probably push a fix tomorrow

**Comment by martmists** [comment]
@gdude2002 @1mn

**Comment by 1mn** [comment]
That's a lot better. Thanks!

**Comment by gdude2002** [comment]
This is a considerable improvement. Good work.
